watcher/tool/watcher/alerts.go

```diff
@@ -0,0 +1,246 @@
+/*
+Copyright 2017 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"strconv"
+
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/pkg/api/v1"
+
+	"github.com/gravitational/monitoring-app/watcher/lib/constants"
+	"github.com/gravitational/monitoring-app/watcher/lib/kapacitor"
+	"github.com/gravitational/monitoring-app/watcher/lib/kubernetes"
+
+	"github.com/ghodss/yaml"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeapi "k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+func runAlertsWatcher(kubernetesClient *kubernetes.Client) error {
+	kapacitorClient, err := kapacitor.NewClient()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	alertLabel, err := kubernetes.MatchLabel(constants.MonitoringLabel, constants.MonitoringUpdateAlert)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	targetLabel, err := kubernetes.MatchLabel(constants.MonitoringLabel, constants.MonitoringUpdateAlertTarget)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	smtpLabel, err := kubernetes.MatchLabel(constants.MonitoringLabel, constants.MonitoringUpdateSMTP)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	alertCh := make(chan map[string]string)
+	alertTargetCh := make(chan map[string]string)
+	configmaps := []kubernetes.ConfigMap{
+		{alertLabel, alertCh},
+		{targetLabel, alertTargetCh},
+	}
+	smtpCh := make(chan map[string][]byte)
+
+	go kubernetesClient.WatchConfigMaps(context.TODO(), configmaps...)
+	go kubernetesClient.WatchSecrets(context.TODO(), kubernetes.Secret{smtpLabel, smtpCh})
+	receiverLoop(context.TODO(), kubernetesClient.Clientset, kapacitorClient,
+		alertCh, alertTargetCh, smtpCh)
+
+	return nil
+}
+
+func receiverLoop(ctx context.Context, kubeClient *kubeapi.Clientset, kClient *kapacitor.Client,
+	alertCh, alertTargetCh <-chan map[string]string, smtpCh <-chan map[string][]byte) {
+	for {
+		select {
+		case update := <-alertCh:
+			spec := []byte(update[constants.ResourceSpecKey])
+			if err := createAlert(kClient, spec); err != nil {
+				log.Warnf("failed to create alert: %v", trace.DebugReport(err))
```

Comment:
I would add more info to these log entries, e.g. at least which configmap/secret it failed to create from, to simplify potential debugging. If this info is not already available here, screw it.

Reply:
screwing it since no info available atm